text/0182-otlp-remote-parent.md

```diff
@@ -0,0 +1,73 @@
+# Export SpanContext.IsRemote in OTLP
+
+Update OTLP to indicate whether a span's parent is remote.
+
+## Motivation
+
+It is sometimes useful to post-process or visualise only entry-point spans: spans which either have no parent (trace roots), or which have a remote parent.
+For example, the Elastic APM solution highlights entry-point spans (Elastic APM refers to these as "transactions") and surfaces these as top-level operations
+in its user interface.
+
+Currently, the only way entry-point spans can be identified is using (lack of) parent ID, and span kind. Relying on span kind can lead to invalid assumptions,
+particularly with relation to `CONSUMER` messaging spans. Using the [batch receiving example](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/messaging.md#batch-receiving)
+in the messaging semantic conventions, `Span Recv1` should be the only entry point into `Process C`. If we assume `CONSUMER` spans are always entry-point spans,
+then this leads to `Span Proc1` and `Span Proc2` being incorrectly classified as entry-point spans. For messaging spans we might also take into account the
```

Comment:
Why incorrectly? `Span Proc1` and `Span Proc2` essentially participate in distributed operation (in this case via links, not parent but anyway). If span context is propagated with consumed messages and you want to visualise that trace, then these spans will be effectively "local roots" or "entry points".

Reply:
> (in this case via links, not parent but anyway)

Just to be clear, I'm talking about the "Batch receiving" example and not "Batch processing". In the former example, the "process" spans are children of the "receive" parent, which is part of the same process.

I meant incorrect in that `Span Proc1` and `Span Proc2` would both considered "local roots", but also have process-local parent -- these are in conflict.

Reply:
@iNikem did that clear things up? Any concerns still, or can this be resolved?

Reply:
It is not clear to me.

What is the desirable outcome? Which of all these spans you want to be classified as "entry-point"?

Reply:
This approach can only classify "entry spans" according to one parent definition. If there are multiple (regular parent, links) then the approach struggles. It would have to be clarified that this "remote parent" information only covers regular parent information.

Reply:
I still don't understand. Can you please list exactly which spans from the referenced "Batch receiving" example are desirable to be classified as "entry-point"? I don't understand what the goal is here.

Reply:
@tigrannajaryan sorry, I'll try to restate in more detail here. Once we've come to a shared understanding I'll update the text.

> Can you please list exactly which spans from the referenced "Batch receiving" example are desirable to be classified as "entry-point"?

`Span Prod1`, `Span Prod2`, and `Span Recv1`

Typically I would expect to see producer spans as children of other span (like a background task, or an HTTP request), in which case `Span Prod1` and `Span Prod2` would no longer be classified as entry-point. They're "entry-point" by definition in the OTEP text because they're trace roots.

More generally: the goal is to identify the spans which represent a request that is entering a service, or originating within a service.

Examples of entry point spans:
 - an incoming HTTP request (entering a service)
 - a message is received from a queue (entering a service)
 - a background task (originating within a service -- trace root)

Examples of non-entry point spans:
 - an outgoing HTTP request
 - a database query
 - message processing (comes _after_ receiving; the receiver span is the entry-point to the service)

Put another way: a trace is a DAG, using only span ID / parent ID (i.e. ignoring links); each service forms a subgraph, and each subgraph has its own [local root](https://github.com/open-telemetry/oteps/pull/182/files#r785611800), or entry point.

The "Batch receiving" example has three traces: Span_Prod1, Span_Prod2, and Span_Recv1+Span_Proc1+Span_Proc3. I've grouped the spans by service/subgraph, and added a border to the local root/entry-point spans:

```mermaid
graph TD
    subgraph producer1
    Span_Prod1
    end

    subgraph producer2
    Span_Prod2
    end

    subgraph batch_receiver
    Span_Recv1 --> Span_Proc1
    Span_Recv1 --> Span_Proc2
    Span_Proc1
    Span_Proc2
    end

    style Span_Prod1 stroke-width:4
    style Span_Prod2 stroke-width:4
    style Span_Recv1 stroke-width:4
```

This is probably not the most illustrative example, and I may have made things confusing by focusing on an edge case. Here's an alternative, but still contrived, example with non-batch receiving messaging spans:

```mermaid
graph TD
    subgraph comments_service
    POST_comments(POST /comment)
    POST_comments --> comments_send(comments send)
    end

    subgraph auth_service
    POST_comments --> POST_auth(POST /auth)
    POST_auth --> LDAP
    end

    subgraph user_details_service
    POST_comments --> GET_user_details(GET /user_details)
    GET_user_details --> SELECT_users(SELECT FROM users)
    end
    
    subgraph comments_inserter
    comments_send --> comments_receive(comments receive)
    comments_receive --> comments_process(comments process)
    comments_process --> INSERT_comments(INSERT INTO comments)
    end

    style POST_comments stroke-width:4
    style POST_auth stroke-width:4
    style GET_user_details stroke-width:4
    style comments_receive stroke-width:4
```

The reason why I focused on the "bulk receiving" example is because we can use other properties to identify entry-points for the more common traces:
 - trace roots are of course trivial to identify
 - server-side request spans can be identified by SERVER span kind
 - by assuming that a CONSUMER span is always an entry point, we can classify the "comments receive" span as an entry point

That assumption is the bit that we want to get rid of, because it's obviously wrong in some cases. Relying on span kind also feels a little brittle, as instrumentation libraries may make a mistake and mess things up.

Reply:
>Span Prod1, Span Prod2, and Span Recv1

All 3 of these spans are parentless. Unless I am missing something the fact that they are parentless is sufficient to classify them as "entry-point". Why is there a need for additional `IsRemote` flag? I think this example hurts more than it helps. It does not show why `IsRemote` flag is justified.

>Here's an alternative, but still contrived, example with non-batch receiving messaging spans

This is indeed what I was missing. It would be great to have this example in the OTEP instead of the Batch Receiving example which confuses me.

>the goal is to identify the spans which represent a request that is entering a service, or originating within a service.

Can we have this in the OTEP text? This seems like **the** goal yet not mentioned so clearly in the OTEP.

--
 
Question 1: do we have the means for instrumentation to easily identify that the span that they are creating has a remote parent? Is it easy to populate the `IsRemote` flag? The OTEP says how the flag will be used, how it will be recorded in the protocol, but it does not say how the instrumentations will be able to correctly set the value of this flag.

Question 2: why can't the backend identify entry points without the `IsRemote` flag? In the example shown the span can be classified as having a remote parent for example if its parent's `service.name` is different. Can a heuristic like this be sufficient to achieve the goal without adding a new flag?

Reply:
Apologies for the delay in responding. I have updated the text and replaced the example.

> Question 1: do we have the means for instrumentation to easily identify that the span that they are creating has a remote parent? Is it easy to populate the IsRemote flag? The OTEP says how the flag will be used, how it will be recorded in the protocol, but it does not say how the instrumentations will be able to correctly set the value of this flag.

Instrumentation does not need to do anything, it's already part of the API and SDKs. From the "Internal details" section:

[`SpanContext.IsRemote`](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md#isremote) identifies whether span context has been propagated from a remote parent.
The OTLP exporter in each SDK would need to be updated to record this in the new `parent_span_is_remote` field.

> Question 2: why can't the backend identify entry points without the IsRemote flag? In the example shown the span can be classified as having a remote parent for example if its parent's service.name is different. Can a heuristic like this be sufficient to achieve the goal without adding a new flag?

That would work, but assumes the backend can see all spans at the time it needs to make the determination whether a span is an entry-point or not; ours cannot. This was also addressed in https://github.com/open-telemetry/oteps/pull/182#discussion_r812340831, please let me know if I should elaborate further.

Reply:
Marking this resolved, please let me know if you still have any doubts about this.